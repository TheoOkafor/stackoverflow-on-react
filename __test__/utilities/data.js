const data = {
  answer: {
    accepted: null,
    body: 'What?',
    comments: [
      {
        answerid: 17,
        body: 'Jonzing',
        id: 30,
        questionid: 4,
        timesubmitted: '2018-11-21T12:35:08.000Z',
        userid: 4,
        username: 'teddy01'
      },
    ],
    downvotes: 0,
    id: 74,
    questionid: 38,
    timesubmitted: '2018-11-21T18:20:11.000Z',
    upvotes: 0,
    userid: 4,
    username: 'teddy01',
    votesCount: 0,
  },
  comment: {
    answerid: 17,
    body: 'Jonzing',
    id: 30,
    questionid: 4,
    timesubmitted: '2018-11-21T12:35:08.000Z',
    userid: 4,
    username: 'teddy01'
  },
  sampleAnswer: {
    data: {
      body: 'Jonzing',
      questionid: 4,
      timesubmitted: '2018-11-21T12:35:08.000Z',
      userid: 4,
      username: 'teddy01'
    },
  },
  question: {
    data: {
      answers: [
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 4,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 1,
          userid: 3,
          username: 'chanpion',
          votesCount: 1
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 8,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 12,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 16,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [{
            answerid: 17,
            body: 'Jonzing',
            id: 30,
            questionid: 4,
            timesubmitted: '2018-11-21T12:35:08.000Z',
            userid: 4,
            username: 'teddy01'
          }],
          downvotes: 0,
          id: 17,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: null,
          body: 'worddy',
          comments: [],
          downvotes: 0,
          id: 67,
          questionid: 4,
          timesubmitted: '2018-11-21T12:23:12.000Z',
          upvotes: 0,
          userid: 4,
          username: 'teddy01',
          votesCount: 0
        }
      ],
      body: 'Lorem ipsum dolor sit amet, ',
      id: 4,
      timesubmitted: '2018-08-22T11:13:09.000Z',
      title: 'How can one learn it all so fast?',
      userid: 3,
      username: 'chanpion'
    },
    message: 'Question 4 found',
    statusCode: 200
  },
  questions: [
    {
      id: 1,
      title: 'why do yo shake?',
      body: 'Lorem ipsum dolor sit amet, ',
      userid: 3,
      username: 'chanpion',
      answers: [],
    },
    {
      id: 2,
      title: 'why do yo shake?',
      body: 'Lorem ipsum dolor sit amet, ',
      userid: 3,
      username: 'chanpion',
      answers: [],
    },
    {
      id: 3,
      title: 'why do yo shake?',
      body: 'Lorem ipsum dolor sit amet, ',
      userid: 3,
      username: 'chanpion',
      answers: [],
    },
  ],
  newQuestion: {
    id: 1,
    title: 'why do yo shake?',
    body: 'Lorem ipsum dolor sit amet, ',
    userid: 3,
    username: 'chanpion',
    answers: [],
  },
  payload: {
    username: 'Xtreme',
    'x-access-token': 'token',
  },
  userData: {
    firstName: 'Xtreme',
    lastName: 'Prodigy',
    username: 'Xtreme',
    password: 'password',
    confirmpassword: 'password',
  },
};

export default data;
