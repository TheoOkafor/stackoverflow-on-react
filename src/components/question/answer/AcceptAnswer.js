import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AcceptAnswer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      data: [],
    };
    this.handleAcceptance = this.handleAcceptance.bind(this);
  }

  handleAcceptance(event) {
    const { id } = event.target;
    let { value } = event.target;
    value = value === 'true' ? value === 'true' : false;
    this.props.handleAccept(value, id);
  }

  render() {
    const user = localStorage.getItem('username');
    const hasRight = user === this.props.username;
    const acceptButton = hasRight && !this.props.hasAccepted
      ? <button
        data-testid="accept"
        className="btn accept right"
        value
        id={this.props.answer.id}
        onClick={this.handleAcceptance}>
        Accept Answer
      </button>
      : null;
    const accepted = this.props.answer.accepted
      ? <div className="accepted right default">Accepted</div>
      : null;
    const unacceptButton = !hasRight || !this.props.answer.accepted
      ? null
      : <button
        data-testid="unaccept"
        className="btn unaccept right show"
        id={this.props.answer.id}
        onClick={this.handleAcceptance}>
          Un-accept Answer
      </button>;
    return (
      <div>
        {unacceptButton}
        {acceptButton}
        {accepted}
      </div>
    );
  }
}

AcceptAnswer.propTypes = {
  id: PropTypes.string,
  index: PropTypes.number,
  answer: PropTypes.object,
  hasAccepted: PropTypes.bool,
  username: PropTypes.string,
  handleAccept: PropTypes.func,
};

export default AcceptAnswer;
