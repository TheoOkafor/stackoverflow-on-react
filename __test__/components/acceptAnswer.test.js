import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, fireEvent, render } from 'react-testing-library';
import { MemoryRouter } from 'react-router-dom';
import AcceptAnswer from '../../src/components/question/answer/AcceptAnswer';
import data from '../utilities/data';

afterEach(cleanup);
localStorage.setItem('username', 'teddy01');
describe('AcceptAnswer component', () => {
  test('Test that the AcceptAnswer displays', () => {
    const component = renderer.create(
      <AcceptAnswer answer={data.answer} index={12}/>
    );
    expect(component).toMatchSnapshot();
  });

  test('should handle click for accept answer', () => {
    const handleClick = jest.fn();
    const { getByTestId } = render(
      <MemoryRouter>
        <AcceptAnswer
          answer={data.answer}
          index={12}
          username="teddy01"
          handleAccept={jest.fn()} />
      </MemoryRouter>
    );
    const acceptBtn = getByTestId('accept');
    acceptBtn.onclick = handleClick;
    fireEvent.click(acceptBtn);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
