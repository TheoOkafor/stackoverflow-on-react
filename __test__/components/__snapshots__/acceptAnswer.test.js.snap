// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcceptAnswer component Test that the AcceptAnswer displays 1`] = `<div />`;
